mflowgen/common/init-fullchip/outputs/alignment-cells.tcl

```diff
@@ -1,58 +1,315 @@
+  # 
+  # EXP 8 (icovlb.6x7-3600y) 6x7, higher still: 6 DTCD errors
+  # gen_fiducial_set [snap_to_grid 1800.00 0.09 99.99] 3600.00 cc true 5 3.0
+}
+
+##############################################################################
+proc gen_fiducial_set {pos_x pos_y {id ul} grid {cols 8} {xsepfactor 1.0}} {
 
-  # Congestion happens at the bottom of the column between like 2200-2700
-  # So let's move them back up, but keep some spacing b/c that was good I think
-  gen_fiducial_set [snap_to_grid 2274.00 0.09 99.99] 2700.00 cc true 0
-  # x,y = 2274,2700
+    # delete_inst -inst ifid_*
+    # set fid_name "init"; # NEVER USED...riiiiiight?
+    # set cols 8
+    
+    # ICOVL cells
+    set width 12.6; # [stevo]: avoid db access by hard-coding width
+    set i_ix_iy [ place_ICOVL_cells $pos_x $pos_y $xsepfactor $id $width $grid $cols ]
+
+    # Unpack return values
+    set i  [ lindex $i_ix_iy 0]
+    set ix [ lindex $i_ix_iy 1]
+    set iy [ lindex $i_ix_iy 2]
+
+    # DTCD cells
+    # There's one feol cell and many beol cells, all stacked in one (ix,iy) place (!!?)
+    set i [ place_DTCD_cell_feol  $i $ix $iy "ifid_dtcd_feol_${id}" $grid ]
+    set i [ place_DTCD_cells_beol $i $ix $iy "ifid_dtcd_beol_${id}"       ]
 }
 
-proc test_vars {} {
-    # Test vars for gen_fiducial_set
-    set pos_x [snap_to_grid 2274.00 0.09 99.99]
-    set pos_y 2700.00
-    set id cc
-    set grid true
-    set cols 0
+proc place_ICOVL_cells { pos_x pos_y xsepfactor id width grid cols } {
+    set i 1;                         # Count how many cells get placed
+    set_dx_dy $id $xsepfactor dx dy; # Set grid x,y spacing
+    set ix $pos_x; set iy $pos_y;    # (pos_x,pos_y) = desired LL corner of grid
+
+    # [stevo]: don't put below/above IO cells
+    set x_bounds [ get_x_bounds $pos_y $grid ]
+
+    foreach cell [ get_ICOVL_cells ] {
+        set fid_name "ifid_icovl_${id}_${i}"
+        create_inst -cell $cell -inst $fid_name \
+            -location "$ix $iy" -orient R0 -physical -status fixed
+        
+        # STYLUS place_inst == LEGACY placeInstance
+        place_inst $fid_name $ix $iy R0 -fixed ; # [stevo]: need this!
+        set ix [ check_pad_overlap $ix $width $x_bounds $grid]
+        # FIXME why do this twice? [stever] I guess in case check_pad_overlap changed $ix??
+        place_inst $fid_name $ix $iy r0; # Overrides/replaces previous placement
+        
+        # Halos and blockages for alignment cells
+        if {$grid == "true"} {
+            set halo_margin_target 15
+        } else {
+            set halo_margin_target 8
+        }
+        set halo_margin [snap_to_grid $halo_margin_target 0.09 0]
+        
+        create_place_halo -insts $fid_name \
+            -halo_deltas $halo_margin $halo_margin $halo_margin $halo_margin -snap_to_site
+        
+        if {$grid == "true"} {
+            create_grid_route_blockages $fid_name $halo_margin
+        } else {
+            create_route_blockage -name $fid_name -inst $fid_name \
```

Comment:
Why is the route blockage smaller for the boundary ICOVL cells? I'm using this in my fullchip flow and I think there are going to be DRCs. I've also had issues using -cover and -spacing for creating blockages in the past. I think we should create identical blockages for all ICOVL cells, regardless of location.

Reply:
All these changes were made because they were able to reduce/prevent/improve actual DRC/routing errors that I saw.  Maybe they will produce new/different routing/DRC errors...but maybe wait until we see those errors before undoing the changes?

Reply:
My recommendation: put sequences like this on some kind of TODO/FIXME/REVISIT-LATER list; if/when we get problems at the end (or not), we can go back and try to clean up things like this to see if it makes a difference...?

Reply:
sounds good